File: ui/src/runner/browser/parser/manifestParser.ts
/**
 * Parse dependency manifests (package.json, go.mod, requirements.txt, pyproject.toml)
 * into structured dependency records. Pure functions — no tree-sitter needed.
 */

export interface ParsedDependency {
  name: string; // "react", "github.com/gorilla/mux", "requests"
  version: string; // "^18.0.0", "v1.8.1", ">=3.0"
  registry: string; // "npm" | "pypi" | "go" | "crates"
  source: string; // manifest file path: "package.json"
  dependencyType: string; // "runtime" | "dev" | "peer" | "optional" | "indirect"
}

export interface ManifestParseResult {
  dependencies: ParsedDependency[];
  errors: string[];
}

const MANIFEST_BASENAMES = new Set([
  'package.json',
  'go.mod',
  'requirements.txt',
  'pyproject.toml',
  'Cargo.toml',
]);

const LOCK_BASENAMES = new Set([
  'package-lock.json',
  'yarn.lock',
  'pnpm-lock.yaml',
  'go.sum',
  'poetry.lock',
  'Cargo.lock',
  'uv.lock',
]);

/** Returns true if the file path is a supported dependency manifest. */
export function isManifestFile(path: string): boolean {
  const basename = path.split('/').pop() ?? '';
  if (LOCK_BASENAMES.has(basename)) return false;
  return MANIFEST_BASENAMES.has(basename);
}

/** Dispatch to the correct parser based on filename. */
export function parseManifest(
  path: string,
  content: string,
): ManifestParseResult {
  const basename = path.split('/').pop() ?? '';
  switch (basename) {
    case 'package.json':
      return parsePackageJson(content, path);
    case 'go.mod':
      return parseGoMod(content, path);
    case 'requirements.txt':
      return parseRequirementsTxt(content, path);
    case 'pyproject.toml':
      return parsePyprojectToml(content, path);
    case 'Cargo.toml':
      return parseCargoToml(content, path);
    default:
      return {
        dependencies: [],
        errors: [`Unsupported manifest: ${basename}`],
      };
  }
}

/** Deterministic package ID: pkg:{registry}:{name} */
export function packageId(registry: string, name: string): string {
  return `pkg:${registry}:${name}`;
}

/** Return the canonical web URL for a package on its registry. */
export function packageSourceUrl(
  registry: string,
  name: string,
): string | undefined {
  switch (registry) {
    case 'npm':
      return `https://www.npmjs.com/package/${name}`;
    case 'pypi':
      return `https://pypi.org/project/${name}/`;
    case 'go':
      return `https://pkg.go.dev/${name}`;
    case 'crates':
      return `https://crates.io/crates/${name}`;
    default:
      return undefined;
  }
}

// ---------------------------------------------------------------------------
// package.json
// ---------------------------------------------------------------------------

export function parsePackageJson(
  content: string,
  source: string,
): ManifestParseResult {
  const errors: string[] = [];
  const deps: ParsedDependency[] = [];

  let parsed: Record<string, unknown>;
  try {
    parsed = JSON.parse(content);
  } catch (e) {
    return { dependencies: [], errors: [`Failed to parse ${source}: ${e}`] };
  }

  const sections: Array<[string, string]> = [
    ['dependencies', 'runtime'],
    ['devDependencies', 'dev'],
    ['peerDependencies', 'peer'],
    ['optionalDependencies', 'optional'],
  ];

  for (const [key, depType] of sections) {
    const section = parsed[key];
    if (!section || typeof section !== 'object') continue;
    for (const [name, version] of Object.entries(
      section as Record<string, string>,
    )) {
      deps.push({
        name,
        version: typeof version === 'string' ? version : '',
        registry: 'npm',
        source,
        dependencyType: depType,
      });
    }
  }

  return { dependencies: deps, errors };
}

// ---------------------------------------------------------------------------
// go.mod
// ---------------------------------------------------------------------------

export function parseGoMod(
  content: string,
  source: string,
): ManifestParseResult {
  const deps: ParsedDependency[] = [];
  const errors: string[] = [];

  // Match require blocks: require ( ... )
  const blockRe = /require\s*\(([\s\S]*?)\)/g;
  let match: RegExpExecArray | null;

  while ((match = blockRe.exec(content)) !== null) {
    const block = match[1];
    for (const line of block.split('\n')) {
      const dep = parseGoRequireLine(line, source);
      if (dep) deps.push(dep);
    }
  }

  // Match single-line requires: require path version
  const singleRe = /^require\s+(\S+)\s+(\S+)(.*)$/gm;
  while ((match = singleRe.exec(content)) !== null) {
    const depType = match[3]?.includes('// indirect') ? 'indirect' : 'runtime';
    deps.push({
      name: match[1],
      version: match[2],
      registry: 'go',
      source,
      dependencyType: depType,
    });
  }

  return { dependencies: deps, errors };
}

function parseGoRequireLine(
  line: string,
  source: string,
): ParsedDependency | null {
  const trimmed = line.trim();
  if (!trimmed || trimmed.startsWith('//')) return null;

  // e.g. "github.com/gorilla/mux v1.8.1 // indirect"
  const parts = trimmed.split(/\s+/);
  if (parts.length < 2) return null;

  const depType = trimmed.includes('// indirect') ? 'indirect' : 'runtime';
  return {
    name: parts[0],
    version: parts[1],
    registry: 'go',
    source,
    dependencyType: depType,
  };
}

// ---------------------------------------------------------------------------
// requirements.txt
// ---------------------------------------------------------------------------

export function parseRequirementsTxt(
  content: string,
  source: string,
): ManifestParseResult {
  const deps: ParsedDependency[] = [];
  const errors: string[] = [];

  for (const rawLine of content.split('\n')) {
    const line = rawLine.trim();
    // Skip blanks, comments, flags (-r, -e, -c, --), and URL references
    if (
      !line ||
      line.startsWith('#') ||
      line.startsWith('-') ||
      line.startsWith('http')
    )
      continue;

    // Parse: name[extras]<op>version or bare name
    const match = line.match(
      /^([A-Za-z0-9_.-]+)(?:\[.*?\])?\s*([<>=!~]+\s*\S+)?/,
    );
    if (!match) continue;

    const rawName = match[1];
    const version = match[2]?.trim() ?? '*';

    deps.push({
      name: normalizePyName(rawName),
      version,
      registry: 'pypi',
      source,
      dependencyType: 'runtime',
    });
  }

  return { dependencies: deps, errors };
}

// ---------------------------------------------------------------------------
// pyproject.toml (limited line-based parser — no TOML library)
// ---------------------------------------------------------------------------

/**
 * Parses dependencies from pyproject.toml using a simple line-based state machine.
 * Handles `[project] dependencies = [...]` and `[tool.poetry.dependencies]` sections.
 * Falls back gracefully on unusual formatting.
 */
export function parsePyprojectToml(
  content: string,
  source: string,
): ManifestParseResult {
  const deps: ParsedDependency[] = [];
  const errors: string[] = [];

  const lines = content.split('\n');
  let i = 0;

  // Track current section header
  let currentSection = '';

  while (i < lines.length) {
    const line = lines[i].trim();

    // Section headers
    if (line.startsWith('[')) {
      currentSection = line;
      i++;
      continue;
    }

    // [project] section: look for dependencies = [...]
    if (currentSection === '[project]' && line.startsWith('dependencies')) {
      const eqIdx = line.indexOf('=');
      if (eqIdx === -1) {
        i++;
        continue;
      }

      const afterEq = line.slice(eqIdx + 1).trim();
      if (afterEq.startsWith('[')) {
        // Could be single-line or multi-line array
        const arrayContent = collectTomlArray(lines, i, eqIdx + 1);
        for (const item of parseTomlStringArray(arrayContent)) {
          const dep = parsePep508(item, source);
          if (dep) deps.push(dep);
        }
      }
      i++;
      continue;
    }

    // [tool.poetry.dependencies] section
    if (currentSection === '[tool.poetry.dependencies]') {
      // key = "version" or key = {version = "...", ...}
      const eqIdx = line.indexOf('=');
      if (eqIdx > 0 && !line.startsWith('#') && !line.startsWith('[')) {
        const name = line.slice(0, eqIdx).trim();
        if (name === 'python') {
          i++;
          continue;
        } // skip python version constraint
        const value = line.slice(eqIdx + 1).trim();
        const version = value.replace(/^["']|["']$/g, '');
        deps.push({
          name: normalizePyName(name),
          version,
          registry: 'pypi',
          source,
          dependencyType: 'runtime',
        });
      }
    }

    // [project.optional-dependencies] section
    if (currentSection.startsWith('[project.optional-dependencies')) {
      if (line.includes('=') && !line.startsWith('#')) {
        const eqIdx = line.indexOf('=');
        const afterEq = line.slice(eqIdx + 1).trim();
        if (afterEq.startsWith('[')) {
          const arrayContent = collectTomlArray(lines, i, eqIdx + 1);
          for (const item of parseTomlStringArray(arrayContent)) {
            const dep = parsePep508(item, source, 'optional');
            if (dep) deps.push(dep);
          }
        }
      }
    }

    i++;
  }

  return { dependencies: deps, errors };
}

/** Collect a TOML array that may span multiple lines, starting from a [ character. */
function collectTomlArray(
  lines: string[],
  startLine: number,
  charOffset: number,
): string {
  let result = lines[startLine].slice(charOffset).trim();
  // Check if the array closes on the same line (track bracket depth)
  if (bracketsClosed(result)) return result;

  for (let i = startLine + 1; i < lines.length; i++) {
    result += '\n' + lines[i];
    if (bracketsClosed(result)) break;
  }
  return result;
}

/** Returns true if all brackets in the string are balanced (outermost [ has its matching ]). */
function bracketsClosed(s: string): boolean {
  let depth = 0;
  for (const ch of s) {
    if (ch === '[') depth++;
    else if (ch === ']') depth--;
    if (depth === 0 && s.indexOf('[') !== -1) return true;
  }
  return false;
}

/** Parse a TOML string array like `["foo>=1.0", "bar"]` into its elements. */
function parseTomlStringArray(raw: string): string[] {
  const items: string[] = [];
  // Match quoted strings inside brackets
  const re = /["']([^"']+)["']/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(raw)) !== null) {
    items.push(m[1]);
  }
  return items;
}

/** Parse a PEP 508 dependency string like "requests>=2.0" or "Flask[async]~=2.0". */
function parsePep508(
  spec: string,
  source: string,
  depType = 'runtime',
): ParsedDependency | null {
  const trimmed = spec.trim();
  if (!trimmed) return null;

  // name[extras]<op>version
  const match = trimmed.match(
    /^([A-Za-z0-9_.-]+)(?:\[.*?\])?\s*([<>=!~]+.*)?$/,
  );
  if (!match) return null;

  return {
    name: normalizePyName(match[1]),
    version: match[2]?.trim() ?? '*',
    registry: 'pypi',
    source,
    dependencyType: depType,
  };
}

/** Normalize Python package name: lowercase, underscores to dashes. */
export function normalizePyName(name: string): string {
  return name.toLowerCase().replace(/_/g, '-');
}

// ---------------------------------------------------------------------------
// Cargo.toml (line-based parser — no TOML library)
// ---------------------------------------------------------------------------

/**
 * Parses dependencies from Cargo.toml using a line-based state machine.
 * Handles [dependencies], [dev-dependencies], and [build-dependencies] sections.
 */
export function parseCargoToml(
  content: string,
  source: string,
): ManifestParseResult {
  const deps: ParsedDependency[] = [];
  const errors: string[] = [];

  const lines = content.split('\n');
  let currentSection = '';

  const depSections: Record<string, string> = {
    '[dependencies]': 'runtime',
    '[dev-dependencies]': 'dev',
    '[build-dependencies]': 'dev',
  };

  for (const rawLine of lines) {
    const line = rawLine.trim();

    // Section headers
    if (line.startsWith('[')) {
      currentSection = line;
      continue;
    }

    if (line.startsWith('#') || !line) continue;

    // Check if we're in a dependency section
    const depType = depSections[currentSection];
    if (!depType) continue;

    const eqIdx = line.indexOf('=');
    if (eqIdx <= 0) continue;

    const name = line.slice(0, eqIdx).trim();
    const value = line.slice(eqIdx + 1).trim();

    // value can be: "1.0", { version = "1.0", ... }, or a bare version
    let version = '*';
    if (value.startsWith('"') || value.startsWith("'")) {
      version = value.replace(/^["']|["']$/g, '');
    } else if (value.startsWith('{')) {
      const verMatch = value.match(/version\s*=\s*["']([^"']+)["']/);
      if (verMatch) version = verMatch[1];
    }

    deps.push({
      name,
      version,
      registry: 'crates',
      source,
      dependencyType: depType,
    });
  }

  return { dependencies: deps, errors };
}

/** Extract npm package name from import specifier: @scope/pkg/sub → @scope/pkg, lodash/fp → lodash */
export function npmPackageName(specifier: string): string {
  if (specifier.startsWith('@')) {
    // Scoped: @scope/pkg or @scope/pkg/sub
    const parts = specifier.split('/');
    return parts.length >= 2 ? `${parts[0]}/${parts[1]}` : specifier;
  }
  // Unscoped: lodash or lodash/fp
  return specifier.split('/')[0];
}

File: agent/src/opentrace_agent/sources/code/manifest_parser.py
# Copyright 2026 OpenTrace Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Manifest file parsers — extract dependency declarations from package manifests.

Ported from the UI's ``manifestParser.ts`` to produce identical output.
Supports: package.json, go.mod, requirements.txt, pyproject.toml, Cargo.toml.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field


@dataclass
class ParsedDependency:
    """A single dependency declaration from a manifest file."""

    name: str
    version: str
    registry: str  # "npm", "pypi", "go", "crates"
    source: str  # manifest filename
    dependency_type: str  # "runtime", "dev", "peer", "optional", "indirect"


@dataclass
class ManifestParseResult:
    """Result of parsing a manifest file."""

    dependencies: list[ParsedDependency] = field(default_factory=list)
    errors: list[str] = field(default_factory=list)


_MANIFEST_NAMES: frozenset[str] = frozenset(
    {
        "package.json",
        "go.mod",
        "requirements.txt",
        "pyproject.toml",
        "Cargo.toml",
    }
)

_LOCK_NAMES: frozenset[str] = frozenset(
    {
        "package-lock.json",
        "yarn.lock",
        "pnpm-lock.yaml",
        "go.sum",
        "poetry.lock",
        "Cargo.lock",
        "uv.lock",
    }
)


def is_manifest_file(filename: str) -> bool:
    """Check whether *filename* (basename only) is a recognised manifest."""
    basename = filename.rsplit("/", 1)[-1]
    if basename in _LOCK_NAMES:
        return False
    return basename in _MANIFEST_NAMES


def normalize_py_name(name: str) -> str:
    """Normalize a Python package name: lowercase, underscores → dashes."""
    return re.sub(r"[-_.]+", "-", name).lower()


def parse_manifest(filename: str, content: str) -> list[ParsedDependency]:
    """Dispatch to the right parser based on *filename*.

    Returns a flat list of dependencies for backward compatibility.
    Use :func:`parse_manifest_result` if you also want error tracking.
    """
    return parse_manifest_result(filename, content).dependencies


def parse_manifest_result(filename: str, content: str) -> ManifestParseResult:
    """Dispatch to the right parser based on *filename*, returning errors too."""
    basename = filename.rsplit("/", 1)[-1]
    parsers = {
        "package.json": parse_package_json,
        "go.mod": parse_go_mod,
        "requirements.txt": parse_requirements_txt,
        "pyproject.toml": parse_pyproject_toml,
        "Cargo.toml": parse_cargo_toml,
    }
    parser = parsers.get(basename)
    if parser is None:
        return ManifestParseResult()
    try:
        deps = parser(content, filename)
        return ManifestParseResult(dependencies=deps)
    except Exception as exc:  # noqa: BLE001
        return ManifestParseResult(errors=[f"{filename}: {exc}"])


# ---------------------------------------------------------------------------
# package.json
# ---------------------------------------------------------------------------

_NPM_SECTIONS = {
    "dependencies": "runtime",
    "devDependencies": "dev",
    "peerDependencies": "peer",
    "optionalDependencies": "optional",
}


def parse_package_json(content: str, source: str) -> list[ParsedDependency]:
    try:
        data = json.loads(content)
    except (json.JSONDecodeError, ValueError):
        return []
    deps: list[ParsedDependency] = []
    for section, dep_type in _NPM_SECTIONS.items():
        section_data = data.get(section)
        if not isinstance(section_data, dict):
            continue
        for name, version in section_data.items():
            deps.append(
                ParsedDependency(
                    name=name,
                    version=str(version),
                    registry="npm",
                    source=source,
                    dependency_type=dep_type,
                )
            )
    return deps


# ---------------------------------------------------------------------------
# go.mod
# ---------------------------------------------------------------------------

_GO_REQUIRE_SINGLE = re.compile(
    r'^\s*require\s+"?([^\s"]+)"?\s+(\S+)',
)
_GO_REQUIRE_BLOCK_LINE = re.compile(
    r"^\s*([^\s/]+)\s+(\S+)",
)


def parse_go_mod(content: str, source: str) -> list[ParsedDependency]:
    deps: list[ParsedDependency] = []
    in_require = False

    for line in content.splitlines():
        stripped = line.strip()

        # Extract module path
        if stripped.startswith("module "):
            continue

        # Single-line require
        m = _GO_REQUIRE_SINGLE.match(stripped)
        if m:
            deps.append(
                ParsedDependency(
                    name=m.group(1),
                    version=m.group(2),
                    registry="go",
                    source=source,
                    dependency_type="runtime",
                )
            )
            continue

        # Block require
        if stripped.startswith("require (") or stripped == "require (":
            in_require = True
            continue
        if in_require:
            if stripped == ")":
                in_require = False
                continue
            m = _GO_REQUIRE_BLOCK_LINE.match(stripped)
            if m:
                dep_type = "indirect" if "// indirect" in line else "runtime"
                deps.append(
                    ParsedDependency(
                        name=m.group(1),
                        version=m.group(2),
                        registry="go",
                        source=source,
                        dependency_type=dep_type,
                    )
                )

    return deps


def extract_go_module_path(content: str) -> str | None:
    """Extract the module path from go.mod content."""
    for line in content.splitlines():
        stripped = line.strip()
        if stripped.startswith("module "):
            return stripped.split(None, 1)[1].strip()
    return None


# ---------------------------------------------------------------------------
# requirements.txt
# ---------------------------------------------------------------------------

_REQ_LINE = re.compile(
    r"^([A-Za-z0-9_.-]+)(?:\[.*?\])?\s*([<>=!~]+\s*\S+)?",
)


def parse_requirements_txt(content: str, source: str) -> list[ParsedDependency]:
    deps: list[ParsedDependency] = []
    for line in content.splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#") or stripped.startswith("-"):
            continue
        m = _REQ_LINE.match(stripped)
        if m:
            name = normalize_py_name(m.group(1))
            version_part = m.group(2)
            if version_part:
                version = re.sub(r"^[<>=!~]+\s*", "", version_part).strip()
            else:
                version = "*"
            deps.append(
                ParsedDependency(
                    name=name,
                    version=version,
                    registry="pypi",
                    source=source,
                    dependency_type="runtime",
                )
            )
    return deps


# ---------------------------------------------------------------------------
# pyproject.toml  (line-based state machine — no TOML library needed)
# ---------------------------------------------------------------------------


def _collect_toml_array(lines: list[str], start_idx: int) -> tuple[list[str], int]:
    """Collect a TOML array that may span multiple lines.

    Starting from the line *after* the opening ``[``, reads lines until brackets
    are balanced.  Returns (collected_items, next_line_index).
    """
    items: list[str] = []
    depth = 1  # we've already seen the opening [
    idx = start_idx

    while idx < len(lines) and depth > 0:
        line = lines[idx].strip()
        for ch in line:
            if ch == "[":
                depth += 1
            elif ch == "]":
                depth -= 1
                if depth == 0:
                    break
        # Extract quoted strings from this line
        items.extend(re.findall(r'"([^"]+)"', line))
        items.extend(re.findall(r"'([^']+)'", line))
        idx += 1

    return items, idx


def parse_pyproject_toml(content: str, source: str) -> list[ParsedDependency]:
    deps: list[ParsedDependency] = []
    lines = content.splitlines()
    section: str | None = None
    i = 0

    while i < len(lines):
        stripped = lines[i].strip()

        # Section headers
        if stripped.startswith("["):
            section = stripped.strip("[]").strip()
            i += 1
            continue

        if section == "project.dependencies":
            # Array entry: "requests>=2.0"
            if stripped.startswith('"') or stripped.startswith("'"):
                dep = _parse_pep508_line(stripped.strip("\"', "))
                if dep:
                    deps.append(
                        ParsedDependency(
                            name=normalize_py_name(dep[0]),
                            version=dep[1],
                            registry="pypi",
                            source=source,
                            dependency_type="runtime",
                        )
                    )
        elif section == "project":
            # Inline or multi-line: dependencies = ["requests>=2.0", ...]
            if stripped.startswith("dependencies") and "=" in stripped:
                after_eq = stripped.split("=", 1)[1].strip()
                if "[" in after_eq:
                    # Check if the array closes on this line
                    if "]" in after_eq:
                        items = re.findall(r'"([^"]+)"', after_eq)
                    else:
                        # Multi-line array
                        items, i = _collect_toml_array(lines, i + 1)
                        # Don't increment i again at the end
                        for item in items:
                            dep = _parse_pep508_line(item)
                            if dep:
                                deps.append(
                                    ParsedDependency(
                                        name=normalize_py_name(dep[0]),
                                        version=dep[1],
                                        registry="pypi",
                                        source=source,
                                        dependency_type="runtime",
                                    )
                                )
                        continue
                    for item in items:
                        dep = _parse_pep508_line(item)
                        if dep:
                            deps.append(
                                ParsedDependency(
                                    name=normalize_py_name(dep[0]),
                                    version=dep[1],
                                    registry="pypi",
                                    source=source,
                                    dependency_type="runtime",
                                )
                            )

        elif section and section.startswith("project.optional-dependencies"):
            if stripped.startswith('"') or stripped.startswith("'"):
                dep = _parse_pep508_line(stripped.strip("\"', "))
                if dep:
                    deps.append(
                        ParsedDependency(
                            name=normalize_py_name(dep[0]),
                            version=dep[1],
                            registry="pypi",
                            source=source,
                            dependency_type="optional",
                        )
                    )

        elif section == "tool.poetry.dependencies":
            # Poetry: key = "version" or key = {version = "..."}
            if "=" in stripped and not stripped.startswith("#"):
                name, _, value = stripped.partition("=")
                name = name.strip().strip('"')
                if name == "python":
                    i += 1
                    continue
                value = value.strip()
                if value.startswith("{"):
                    vm = re.search(r'version\s*=\s*"([^"]*)"', value)
                    version = vm.group(1) if vm else "*"
                else:
                    version = value.strip('"')
                deps.append(
                    ParsedDependency(
                        name=normalize_py_name(name),
                        version=version,
                        registry="pypi",
                        source=source,
                        dependency_type="runtime",
                    )
                )

        elif section == "tool.poetry.dev-dependencies":
            if "=" in stripped and not stripped.startswith("#"):
                name, _, value = stripped.partition("=")
                name = name.strip().strip('"')
                value = value.strip()
                if value.startswith("{"):
                    vm = re.search(r'version\s*=\s*"([^"]*)"', value)
                    version = vm.group(1) if vm else "*"
                else:
                    version = value.strip('"')
                deps.append(
                    ParsedDependency(
                        name=normalize_py_name(name),
                        version=version,
                        registry="pypi",
                        source=source,
                        dependency_type="dev",
                    )
                )

        i += 1

    return deps


def _parse_pep508_line(spec: str) -> tuple[str, str] | None:
    """Parse a PEP 508 dependency spec like ``requests>=2.0`` → (name, version)."""
    m = re.match(r"^([A-Za-z0-9][\w.\-]*)", spec)
    if not m:
        return None
    name = m.group(1)
    rest = spec[m.end() :]
    # Strip extras like [security]
    rest = re.sub(r"^\[.*?\]", "", rest)
    version_match = re.search(r"[><=!~]+\s*(\S+)", rest)
    version = version_match.group(1).rstrip(",;") if version_match else "*"
    return name, version


# ---------------------------------------------------------------------------
# Cargo.toml  (line-based)
# ---------------------------------------------------------------------------


def parse_cargo_toml(content: str, source: str) -> list[ParsedDependency]:
    deps: list[ParsedDependency] = []
    section: str | None = None

    for line in content.splitlines():
        stripped = line.strip()

        if stripped.startswith("["):
            section = stripped.strip("[]").strip()
            continue

        if section in ("dependencies", "dev-dependencies", "build-dependencies"):
            dep_type = {
                "dependencies": "runtime",
                "dev-dependencies": "dev",
                "build-dependencies": "dev",
            }[section]
            if "=" in stripped and not stripped.startswith("#"):
                name, _, value = stripped.partition("=")
                name = name.strip()
                value = value.strip().strip('"')
                # Handle inline table: foo = { version = "1.0", features = [...] }
                if value.startswith("{"):
                    vm = re.search(r'version\s*=\s*"([^"]*)"', value)
                    version = vm.group(1) if vm else "*"
                else:
                    version = value
                deps.append(
                    ParsedDependency(
                        name=name,
                        version=version,
                        registry="crates",
                        source=source,
                        dependency_type=dep_type,
                    )
                )

    return deps
